Introduce a ResizableBox component

chrisvanpatten opened the pull request:
This is a follow-up to #10331 and fixes #10343.

This PR…

+ introduces a `ResizableBox` component which wraps the `re-resizable` package to remove some code duplication
+ introduces global/shared classes for resize handles
+ removes the mixins introduced in #10331, and thus the code duplication (which had existed for a while before that)
+ updates `core/image` and `core/spacer` to use this new component

~I plan to flesh out the docs, but want to make sure I'm on the right track with this first!~ Still want to know if I'm on track, but I added additional docs anyway :)

## Testing

+ Insert image and spacer blocks
+ Verify there are no visual or functional regressions at various alignments and configurations

chrisvanpatten committed:
Introduce a ResizableBox component and apply to image/spacer blocks

chrisvanpatten committed:
Add a shared class to all drag handles to simplify CSS selectors

chrisvanpatten committed:
Add additional documentation and an example

tofumatt commented:
Thanks for jumping on the patch for this! 😄

I'l try to take a look this week, but I might have to wait a bit until after we clear out the 4.0 milestone. 😅

tofumatt committed:
chore: Tweak docs

tofumatt reviewed (changes requested):
I dig it! I think we need to add to the `components` changelog to mention the new component we've introduced. 🤔

After that I think this is good! Thanks for doing this! ❤️

chrisvanpatten commented on packages/components/src/resizable-box/README.md:
```diff
@@ -0,0 +1,55 @@
+
+## Usage
+
+Most options are passed directly through to [re-resizable](https://github.com/bokuweb/re-resizable) so you may wish to refer to their documentation. Note that we set the `handleClasses` and `handleStyles` to enable the Gutenberg-specific styles.
```
Fair enough, I can see how it would be slightly confusing… we're _setting_ the prop but _unsetting_ the values. I'll update this :)

chrisvanpatten committed:
Add changelog comment and update readme

chrisvanpatten commented:
@tofumatt Clarified the docs and added to the changelog. Might need a check from @gziolo to be sure I updated the changelog correctly (although I do believe it's right; I checked the file history to see how it had been handled previously).

chrisvanpatten committed:
Removed a stray "the"

tofumatt commented on packages/components/CHANGELOG.md:
```diff
@@ -1,3 +1,9 @@
+## 5.0.0 (Unreleased)
```
Not sure I'd count this as a `5.0.0` feature; it's more like `4.1.0`. It isn't a breaking change and it's just an addition.

chrisvanpatten replied:
Fair point :)

chrisvanpatten committed:
s/5.0.0/4.1.0/g

tofumatt commented:
I think the test failures are unrelated, see: #10364.

Once that's merged you should probably rebase and then flag me for another review... but I think this is all good 👍

chrisvanpatten commented:
@tofumatt looks like Travis is still failing for the Undo and Split/Merge tests, which is likely unrelated.

tofumatt commented:
Yup, that one is an "expected" intermittent failure 😅

tofumatt commented:
It keeps failing, maybe something else is up? 😢

chrisvanpatten commented:
@tofumatt I’ll try the tests locally. It’s strange it only happens in the one environment in Travis…

tofumatt commented:
Ten four, I've kicked the test server again and hopefully it'll work done 🤷‍♂️

chrisvanpatten commented:
Strangely I can't "pass" the undo test in `master` (or this branch) even when I manually follow the instructions (typing in three lines and hitting undo six times). It actually takes me 7 or 8 undos before I'm back to a disabled undo state.

I'm wondering if this might actually be a bug in `master` right now? I can't reproduce it in 3.9.

chrisvanpatten commented:
lol okay just as I type that the test finally passes ¯\_(ツ)_/¯

@tofumatt I think we're finally clear on this… phew…

tofumatt commented:
Miguel and I were saying that actually replicating the test case manually indicates it might be a bit off. I think that test case could use some deeper examination…

- Matt (Sent from mobile)

> On 6 Oct 2018, at 13:28, Chris Van Patten <notifications@github.com> wrote:
> 
> lol okay just as I type that the test finally passes ¯_(ツ)_/¯
> 
> @tofumatt I think we're finally clear on this… phew…
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

tofumatt commented:
In general I’m for merging as much stuff as possible and this is pretty minimal. I’ll have a look tomorrow, been a bit distracted by travel today.

There will be a 4.1 by the way! In general Gutenberg development (including releases) will continue on GitHub, so while I’m not exactly sure how things will work after WordPress 5.0, I think the general idea will be that certain releases of WordPress will pull in a certain release of Gutenberg, if that makes sense.

chrisvanpatten commented:
@tofumatt totally fair! Travel can’t knock the wind out of you.

I’m going to try a git blame to see if I can figure out when the undo behaviour changed and if I can figure it out I’ll open a new issue.

chrisvanpatten commented:
@tofumatt just opened #10380 which tracks the undo issue. It looks related to the rich text state structure changes, which makes sense.

tofumatt commented:
Thanks for tracking that one down!

- Matt (Sent from mobile)

> On 6 Oct 2018, at 22:00, Chris Van Patten <notifications@github.com> wrote:
> 
> @tofumatt just opened #10380 which tracks the undo issue. It looks related to the rich text state structure changes, which makes sense.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

tofumatt reviewed (approved):
I dig it 👍

aduth commented:
This introduced changes when running `npm install` on master because the package-lock.json` file wasn't updated.

aduth commented on packages/components/src/resizable-box/style.scss:
```diff
@@ -0,0 +1,43 @@
+	display: none;
+
+	// Show the resize handle when selected.
+	.components-resizable-box__container.is-selected & {
```
Where do we document how a resizable box becomes "selected" to unhide itself? The current documentation for `ResizableBox` has no indication of this default hidden state. It was quite the journey to find my way here, as I've found it quite difficult to use the component.

chrisvanpatten replied:
The behavior here was inherited from the original, non-componentized usages of ResizableBox; you can see it in previous PRs as well such as [here](https://github.com/WordPress/gutenberg/pull/10331/files#diff-01bcf979fbfe3fb1e688ddcddcbedd66).

There are probably other use-cases that would provide more valuable context, and could certainly be referenced and improved in the documentation.

aduth replied:
While I'd agree it often needs contextualized styling, the main problem for me was that in using the example from the documentation, there is nothing visual which appears. It took me quite a while to realize that the handles are hidden by default.

I understand this is largely a result of how it was used prior to this pull request, but it seems to me that as a generic component, there's no reason to hide those controls by default. I could imagine we might want to keep this behavior for blocks in a block list, but rather than styling in ResizableBox and in individual block stylesheets, it seems like something which should be managed in one of the stylesheets for the block list (apply consistently and exclusively to ResizableBox within blocks).
